netket/optimizer/qgt/qgt_jacobian_pytree_logic.py

```diff
@@ -0,0 +1,226 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Optional, Tuple
+from functools import partial
+
+import jax
+import jax.flatten_util
+import jax.numpy as jnp
+
+import numpy as np
+
+from netket.stats import sum_inplace, subtract_mean
+from netket.utils import n_nodes
+
+from netket.utils.types import Array, Callable, PyTree, Scalar
+
+import netket.jax as nkjax
+from netket.jax import tree_cast, tree_conj, tree_axpy, tree_to_real
+
+
+def sub_mean(oks: PyTree) -> PyTree:
+    return jax.tree_map(partial(subtract_mean, axis=0), oks)  # MPI
+
+
+@partial(jax.vmap, in_axes=(None, None, 0))
+def vmap_grad_real_holo(forward_fn: Callable, params: PyTree, samples: Array) -> PyTree:
+    """Calculates Jacobian entries by vmapping grad.
+    Assumes the function is R→R or holomorphic C→C, so single grad is enough
+
+    Args:
+        forward_fn: the log wavefunction, partialed with any model state, limited to single sample, taking a split real-imaginary PyTree in R→R mode
+        params : a pytree of parameters p, real-imaginary split in R→R mode
+        samples : an array of n samples σ
+
+    Returns:
+        The Jacobian matrix ∂/∂pₖ ln Ψ(σⱼ) as a PyTree
+    """
+    y, vjp_fun = jax.vjp(forward_fn, params, samples)
+    res, _ = vjp_fun(np.array(1.0, dtype=jnp.result_type(y)))
+    return res
+
+
+def vmap_grad_centered_real_holo(
+    forward_fn: Callable, params: PyTree, samples: Array
+) -> PyTree:
+    """Calculates centred Jacobian (i.e., subtracts MPI mean from vmap_grad)"""
+    return sub_mean(vmap_grad_real_holo(forward_fn, params, samples))
+
+
+@partial(jax.vmap, in_axes=(None, None, 0))
+def vmap_grad_cplx(forward_fn: Callable, params: PyTree, samples: Array) -> PyTree:
+    """Calculates Jacobian entries by vmapping grad.
+    Assumes the function is R→C, backpropagates 1 and -1j
+
+    Args:
+        forward_fn: the log wavefunction, partialed with any model state, limited to single sample, taking a split real-imaginary PyTree
+        params : a pytree of parameters p, real-imaginary split
+        samples : an array of n samples σ
+
+    Returns:
+        The Jacobian matrix ∂/∂pₖ ln Ψ(σⱼ) as a PyTree"""
+    y, vjp_fun = jax.vjp(forward_fn, params, samples)
+    gr, _ = vjp_fun(np.array(1.0, dtype=jnp.result_type(y)))
+    gi, _ = vjp_fun(np.array(-1.0j, dtype=jnp.result_type(y)))
+    return gr, gi
+
+
+def vmap_grad_centered_cplx(
+    forward_fn: Callable, params: PyTree, samples: Array
+) -> PyTree:
+    """Calculates centred Jacobian (i.e., subtracts MPI mean from vmap_grad)"""
+    gr, gi = vmap_grad_cplx(forward_fn, params, samples)
+    # Return the real and imaginary parts of ΔOⱼₖ stacked along the sample axis
+    # Re[S] = Re[(ΔOᵣ + i ΔOᵢ)ᴴ(ΔOᵣ + i ΔOᵢ)] = ΔOᵣᵀ ΔOᵣ + ΔOᵢᵀ ΔOᵢ = [ΔOᵣ ΔOᵢ]ᵀ [ΔOᵣ ΔOᵢ]
+    return jax.tree_multimap(
+        lambda re, im: jnp.concatenate([re, im], axis=0), sub_mean(gr), sub_mean(gi)
+    )
```

Comment:
This is the code that calculates the derivatives here, could you please check whether everything looks OK? 
Specifically, I took this bit from @inailuig's original logic without thinking about it at all:
```python
    gr, _ = vjp_fun(np.array(1.0, dtype=jnp.result_type(y)))
    gi, _ = vjp_fun(np.array(-1.0j, dtype=jnp.result_type(y)))
```
Are we sure this gives what we want?